File: app/src/main/java/net/squanchy/eventdetails/widget/EventDetailsLayout.java
package net.squanchy.eventdetails.widget;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.Resources;
import android.os.Build;
import android.support.annotation.AttrRes;
import android.support.annotation.ColorInt;
import android.support.annotation.Nullable;
import android.text.Html;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.style.ForegroundColorSpan;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import net.squanchy.R;
import net.squanchy.schedule.domain.view.Event;
import net.squanchy.schedule.domain.view.Place;
import net.squanchy.support.lang.Optional;

import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

public class EventDetailsLayout extends LinearLayout {

    private static final String WHEN_DATE_TIME_FORMAT = "EEEE, d MMMM 'at' HH:mm";

    private View whenContainer;
    private TextView whenTextView;
    private View whereContainer;
    private TextView whereTextView;
    private View descriptionHeader;
    private TextView descriptionTextView;

    public EventDetailsLayout(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public EventDetailsLayout(Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        super.setOrientation(VERTICAL);
    }

    @Override
    public void setOrientation(int orientation) {
        throw new UnsupportedOperationException("Changing orientation is not supported for EventDetailsLayout");
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        View.inflate(getContext(), R.layout.merge_event_details_layout, this);

        whenTextView = (TextView) findViewById(R.id.when_text);
        whenContainer = findViewById(R.id.when_container);
        whereTextView = (TextView) findViewById(R.id.where_text);
        whereContainer = findViewById(R.id.where_container);
        descriptionHeader = findViewById(R.id.description_header);
        descriptionTextView = (TextView) findViewById(R.id.description_text);
    }

    public void updateWith(Event event) {
        updateWhen(event);
        updateWhere(event);
        updateDescription(event.description());
    }

    private void updateWhen(Event event) {
        DateTimeFormatter formatter = DateTimeFormat.forPattern(WHEN_DATE_TIME_FORMAT)
                .withZone(event.timeZone());
        whenTextView.setText(formatter.print(event.startTime().toDateTime()));
        whenContainer.setVisibility(VISIBLE);
    }

    private void updateWhere(Event event) {
        if (event.place().isPresent()) {
            whereContainer.setVisibility(VISIBLE);
            whereTextView.setText(placeTextFrom(event.place().get()));
        } else {
            whereContainer.setVisibility(GONE);
        }
    }

    private CharSequence placeTextFrom(Place place) {
        SpannableStringBuilder builder = new SpannableStringBuilder(place.name());
        if (place.floor().isPresent()) {
            String floorLabel = place.floor().get();
            builder.append("   ")
                    .append(floorLabel)
                    .setSpan(
                            createColorSpan(whereTextView, android.R.attr.textColorSecondary),
                            builder.length() - floorLabel.length(),
                            builder.length(),
                            Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
                    );
        }
        return builder;
    }

    private ForegroundColorSpan createColorSpan(View targetView, @AttrRes int attributeResId) {
        int color = getColorFromTheme(targetView.getContext().getTheme(), attributeResId);
        return new ForegroundColorSpan(color);
    }

    @ColorInt
    private int getColorFromTheme(Resources.Theme theme, @AttrRes int attributeId) {
        TypedValue typedValue = new TypedValue();
        theme.resolveAttribute(attributeId, typedValue, true);
        return typedValue.data;
    }

    private void updateDescription(Optional<String> description) {
        if (description.isPresent()) {
            descriptionHeader.setVisibility(VISIBLE);
            descriptionTextView.setVisibility(VISIBLE);
            descriptionTextView.setText(parseHtml(description.get()));
        } else {
            descriptionHeader.setVisibility(GONE);
            descriptionTextView.setVisibility(GONE);
        }
    }

    @TargetApi(Build.VERSION_CODES.N)
    @SuppressWarnings("deprecation")        // The older fromHtml() is only called pre-24
    private Spanned parseHtml(String description) {
        // TODO handle this properly
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            return Html.fromHtml(description, Html.FROM_HTML_MODE_LEGACY);
        } else {
            return Html.fromHtml(description);
        }
    }
}

File: app/src/main/java/net/squanchy/eventdetails/widget/EventDetailsLayout.kt
package net.squanchy.eventdetails.widget

import android.content.Context
import android.graphics.Color
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.style.ForegroundColorSpan
import android.util.AttributeSet
import android.view.View
import android.widget.TextView
import androidx.annotation.AttrRes
import androidx.annotation.ColorInt
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.content.ContextCompat
import androidx.core.view.isVisible
import arrow.core.Option
import kotlinx.android.synthetic.main.merge_event_details_layout.view.*
import net.squanchy.R
import net.squanchy.eventdetails.domain.view.ExperienceLevel
import net.squanchy.schedule.domain.view.Event
import net.squanchy.schedule.domain.view.Place
import net.squanchy.schedule.domain.view.Track
import net.squanchy.support.content.res.getColorFromAttribute
import net.squanchy.support.lang.getOrThrow
import net.squanchy.support.text.parseHtml
import org.joda.time.DateTimeZone
import org.joda.time.LocalDateTime
import org.joda.time.format.DateTimeFormat

class EventDetailsLayout @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet?,
    defStyle: Int = 0
) : ConstraintLayout(context, attrs, defStyle) {

    private val dateTimeFormatter = DateTimeFormat.forPattern(WHEN_DATE_TIME_FORMAT)

    override fun onFinishInflate() {
        super.onFinishInflate()

        inflate(context, R.layout.merge_event_details_layout, this)
    }

    fun updateWith(event: Event) {
        updateWhen(event.startTime, event.timeZone)
        updateWhere(event.place)
        updateLevel(event.experienceLevel, event.type)
        updateTrack(event.track)
        updateDescription(event.description)
    }

    private fun updateWhen(startTime: LocalDateTime, timeZone: DateTimeZone) {
        val formatter = dateTimeFormatter.withZone(timeZone)
        whenValue.text = formatter.print(startTime.toDateTime(timeZone))
        whenGroup.isVisible = true
    }

    private fun updateWhere(place: Option<Place>) {
        if (place.isDefined()) {
            whereGroup.isVisible = true
            whereValue.text = place.getOrThrow().toPlaceLabel()
        } else {
            whereGroup.isVisible = false
        }
    }

    private fun Place.toPlaceLabel(): CharSequence {
        val builder = SpannableStringBuilder(name)
        if (floor.isDefined()) {
            val floorLabel = floor.getOrThrow()
            builder.append("   ")
                .append(floorLabel)
                .setSpan(
                    createColorSpan(whereValue, android.R.attr.textColorSecondary),
                    builder.length - floorLabel.length,
                    builder.length,
                    Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
                )
        }
        return builder
    }

    private fun createColorSpan(targetView: View, @AttrRes attributeResId: Int): ForegroundColorSpan {
        val color = targetView.context.theme.getColorFromAttribute(attributeResId)
        return ForegroundColorSpan(color)
    }

    private fun updateLevel(level: Option<ExperienceLevel>, type: Event.Type) {
        when {
            type == Event.Type.KEYNOTE -> {
                levelGroup.isVisible = false
            }
            level.isDefined() -> {
                levelGroup.isVisible = true

                val experienceLevel = level.getOrThrow()
                levelValue.setText(experienceLevel.labelStringResId)

                val experienceColor = ContextCompat.getColor(context, experienceLevel.colorResId)
                levelValue.tintCompoundDrawableEnd(experienceColor)
            }
            else -> levelGroup.isVisible = false
        }
    }

    private fun updateTrack(trackOption: Option<Track>) {
        if (trackOption.isDefined()) {
            trackGroup.isVisible = true

            val track = trackOption.getOrThrow()
            trackValue.text = track.name
            track.accentColor?.let {
                val trackColor = Color.parseColor(it)
                trackValue.tintCompoundDrawableEnd(trackColor)
            }
        } else {
            trackGroup.isVisible = false
        }
    }

    private fun TextView.tintCompoundDrawableEnd(@ColorInt color: Int) {
        val compoundDrawables = compoundDrawablesRelative
        val endCompoundDrawable = compoundDrawables[2]
        endCompoundDrawable?.setTint(color)
    }

    private fun updateDescription(description: Option<String>) {
        if (description.isDefined()) {
            descriptionGroup.isVisible = true
            descriptionValue.text = description.getOrThrow().parseHtml()
        } else {
            descriptionGroup.isVisible = false
        }
    }

    companion object {

        private const val WHEN_DATE_TIME_FORMAT = "EEEE, d MMMM 'at' HH:mm"
    }
}
